### PR TITLE
RFD75: Snowflake Integration - Database Access

### PR DESCRIPTION
**Pull request by jakule**
RFD for Snowflake DB integration. We're trying to confirm some undocumented features with Snowflake to makes sure that we can use them in our implementation. I'll update the RFD when we get a response from them.

### CODE REVIEW

File: rfd/00xx-snowflake-support.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+### Teleport configuration
+
+Like in case of other databases, a user will have to add an entry to the teleport configuration file. Example:
+
+```yaml
+db_service:
+  enabled: "yes"
+  databases:
+      - name: "snowflake"
+      description: "Snowflake database"
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:40]: You forgot the mandatory part of every description, a Unicode symbol :)
```suggestion
    - name: "snowflake"
      description: "❄ Snowflake database"
```

File: rfd/00xx-snowflake-support.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      protocol: "snowflake"
+      uri: "im12345.us-east-2.aws.snowflakecomputing.com"
+```
+
+The next step would be to add teleport CA to the Snowflake user account. `tctl` could learn to export the required keys
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:45]: Are we going to use our new Database CA to sign JWT tokens?

File: rfd/00xx-snowflake-support.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      uri: "im12345.us-east-2.aws.snowflakecomputing.com"
+```
+
+The next step would be to add teleport CA to the Snowflake user account. `tctl` could learn to export the required keys 
+in the same way as it does today for different databases (`tctl auth sign` command). The exported key must then be added 
+to Snowflake by an administrator. The procedure of adding the key to Snowflake is described [here](https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-4-assign-the-public-key-to-a-snowflake-user).
+
+### Database access support
+
+`tsh db login` should export a user private key in PCKS8 format (that's the only format supported by `snowsql` and some SDKs).
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:51]: Is this referring to the user's private key that's kept in `~/.tsh/keys/proxy.example.com/alice` (which is generated during `tsh login`)?

File: rfd/00xx-snowflake-support.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+That key should be used to sign a "Teleport JWT". Client then should send the request with generated JWT to authenticate 
+with Snowflake. Teleport can intercept HTTP communication between a client and the Snowflake. The first request send by 
+the client should be to the `/session/v1/login-request` endpoint which contains all login information as username, JWT. 
+This data can be used to authenticate user in Teleport. JWT token prepared by the client can be then read and replaces by Snowflake JWT
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:55]: This is not super clear. Since this is HTTP, does the client send authentication info (JWT) with every request? What does "authenticate user in Teleport" mean?

File: rfd/00xx-snowflake-support.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    participant Snowflake
+    client->>Teleport: HTTP /session/v1/login-request
+    Teleport->>Teleport: RBAC check, genererate and replace JWT
+    Teleport->>Snowflake: HTTP /session/v1/login-request
+    Snowflake->>Teleport: Authentication Token
+    Teleport->>Teleport: Generate Teleport Auth token
+    Teleport->>client: Send Teleport Auth token
+```
+
+Nn each request Teleport can replace the "Teleport Auth token" with Snowflake Auth token:
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:74]: "Nn each request" <-- feels like a typo

File: rfd/00xx-snowflake-support.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
+the token as client doesn't have access to the token (login response contains the time for how long the token is valid).
+
+#### Query logging
+
+Queries can be extracted from requests sent to `/queries/v1/query-request` endpoint, field `data.sqltext`.
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:95]: Is there protocol documentation we can link here?

File: rfd/00xx-snowflake-support.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+when connecting to the Snowflake on a different port that 443. 
+
+`snowsql` requires to use one of the available authentication method. By default, it asks for password unless other 
+authentication option is specified. The other alternatives are SSO and JWT/Private key (https://docs.snowflake.com/en/user-guide/snowsql-start.html).
+Teleport could use JWT/Private key for authentication. In this flow `tsh` would export user's private in PKCS8 format (at that's the only one supported by `snowsql`),
+and `snowsql` could use it to sign the JWT token.
+
+I think there are two ways on how we can support the `snowsql`: 
+
+1. `snowcli` can connect to a custom host **without** TLS on any port different from 443.
```
Comment (r0mant) [inline rfd/00xx-snowflake-support.md:115]: We'll have to support `tsh proxy db --tunnel` mode where the client connects to the local authenticate tunnel without TLS anyway. I would just opt for that approach for now.

**Review by r0mant — APPROVED** [review]
Just a few questions/comments, otherwise let's implement it!

File: rfd/00xx-snowflake-support.md (hunk 2 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:45]: I think so. We don't need to port the integration to v9 and any other CA at that point will be confusing as all other databases are using the Database CA. Rotation could potentially break the integration.

File: rfd/00xx-snowflake-support.md (hunk 6 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:95]: I wasn't able to find anything official 😞 Maybe Snowflake will point us to some documentation.

File: rfd/00xx-snowflake-support.md (hunk 1 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:40]: Good catch. Reading the document I felt that something was missing. Now I know what was it 😄

File: rfd/00xx-snowflake-support.md (hunk 5 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:74]: Fixed

File: rfd/00xx-snowflake-support.md (hunk 3 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:51]: Yes, added the path.

File: rfd/00xx-snowflake-support.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+Below example shows how to start `snowsql` in that mode:
+
+```shell
+export https_proxy=localhost:2000
+export http_proxy=localhost:2000
+export CURL_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem # includes the Teleport issued CA
+export REQUESTS_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem
+
+snowsql --noup -a im56867.us-east-2.aws -u teleport-user --private-key-path ${HOME}/.tsh/keys/example.com/bob.pkcs8 -h me.localhost
```
Comment (smallinsky) [inline rfd/00xx-snowflake-support.md:132]: Does the snowsql  tools trigger some kind of  HTTP requests that are not stricte related to Snowflake SQL API. For example client upload version check ?. I that case I think that is is good idea to forward the request to the url without `local_proxy->teleport_proxy->app agent` flow.

File: rfd/00xx-snowflake-support.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    participant Snowflake
+    client->>Teleport: HTTP /queries/v1/query-request
+    Teleport->>Teleport: Replace Auth header
+    Teleport->>Snowflake: HTTP /queries/v1/query-request
+    Snowflake->>Teleport: Response
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
```
Comment (smallinsky) [inline rfd/00xx-snowflake-support.md:90]: Wonder about situation where we have 2  or more HA Snowflake agent. Does the API allows to generate and uses `n` tokens simultaneously ? If not we should store and sync token generation between HA Snowflake agents.

File: rfd/00xx-snowflake-support.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
```
Comment (smallinsky) [inline rfd/00xx-snowflake-support.md:91]: Just leaving a remainder comment that we should ensure that Snowflake API doesn't have any Quota for token refreshing operation.

File: rfd/00xx-snowflake-support.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+the token as client doesn't have access to the token (login response contains the time for how long the token is valid).
+
+#### Query logging
+
+Queries can be extracted from requests sent to `/queries/v1/query-request` endpoint, field `data.sqltext`.
+
+#### User authentication
+
+Username can be extracted from `/session/v1/login-request` requests, field `data.LOGIN_NAME`. The extracted username
```
Comment (smallinsky) [inline rfd/00xx-snowflake-support.md:100]: We have also `RouteToDatabase.Username` field in case of TLS connection where the cert can be set on client side so we need to make that cert db username will be also properly validated.

File: rfd/00xx-snowflake-support.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+* environment variable `https_proxy` - sets the address of `tsh proxy db` tunnel address. Traffic will be sent there.
+* environment variable `CURL_CA_BUNDLE` - passes Teleport's CA bundle, so we can terminate TLS connection with our certificate.
+* `--private-key-path` - sets the path to PCKS8 private key that will be used to sign JWT.
+* `-h me.localhost` - is the name of destination server, in our case this name has to match the name on the certificate that we will use for TLS termination.
+
+### UX
+
+Snowflake will integrate with teleport in the same way as other databases.
```
Comment (smallinsky) [inline rfd/00xx-snowflake-support.md:142]: In that case where the Snowflake JWT token will be stored ? In app access we have backend App/WebSession but I don't think that we have  an equivalent of this backend type for database access. 
Generating JWT token on each request can drain to Snowflake quota where storing token in memory at first glance seems to be find but not sure a Snowflake token can generate and use N simultaneous tokens (HA case with meny Snowflake agents)

File: rfd/00xx-snowflake-support.md (hunk 9 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:90]: Good point. I noticed the problem when I started writing the implementation. I think there is a few scenarios related to tokens:
1. Single connection per client application, single DB proxy (ex. some GUI clients).
2. Multiple connections per client application, single DB proxy (ex.`snowsql` uses 4 connections by default).
3. Multiple connections per client application, DB proxy in HA configuration.

We need at least one token per connected user. I think that we can assume that we can get one token per connected client as I don't think that sharing tokens between multiple users is a good idea. API is not documented (at least I wasn't able to find any official docs), so I'm not sure what limits apply (how many tokens can we generate, what is the rate limiting, etc).
The simplest implementation would generate one token per connection and keep it in memory. Connection comes from the client, we generate a new token. That works fine for clients that create only one connection, as client using multiple connections would generate multiple tokens. We would generate more token than we need. Each connection may go to different DB agent, so we would need to share the tokens if we don't want to generate extra tokens. That would complicate the implementation tho. Let's assume that you're using `snowsql`. At the beginning `snowsql` creates one connection to authenticate the client. Right after the successful response is returned `snowsql` creates additional connections using the generated token. Sharing the token would have to be quick, because new connection can go to another DB agent that doesn't have the token generated by different DB agent. We could use some kind of "sticky session" to keep all connection from a single machine routed to the same DB proxy or we could encrypt the access token and return it back to the client. That would simplify the flow on our side. That would also help us with token rotation as client needs to request the rotation anyway, so we could rotate the token only then. We would just need to make sure that the encrypted token returned to a client is secure 😅 

TLDR:
1. Create one token per connection. 
Pros: simple implementation
Cons: we can potentially hit some limits as we don't know what they are. We also need to rotate each token in the background.
2. Return encrypted token to the client
Pros: Should work in any configuration and reduces the number of tokens we generate
Cons: Potential security risk if we implement the encryption wrong.

File: rfd/00xx-snowflake-support.md (hunk 8 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:132]: We could use `NO_PROXY` environment variable, https://docs.snowflake.com/en/user-guide/snowsql-start.html#using-a-proxy-server that that would require to list all domains that should skip the proxy.

File: rfd/00xx-snowflake-support.md (hunk 10 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:91]: I hope that maybe Snowflake will share some docs with us, so we know what the limit is. Until then I can just add, that there maybe some limit 😐

File: rfd/00xx-snowflake-support.md (hunk 11 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:100]: Agree. I forgot to mention that. I'll update the RFD.

File: rfd/00xx-snowflake-support.md (hunk 12 above)
Comment (jakule) [reply rfd/00xx-snowflake-support.md:142]: https://github.com/gravitational/teleport/pull/12006#discussion_r858070934

Comment (jakule) [reply rfd/00xx-snowflake-support.md:142]: One thing that can be not obvious. All clients that I've tested so far keep connection open, so we can just keep the tokens in the memory until the connection ends.

**Review by smallinsky — APPROVED** [review]
